girder/models/model_base.py

```diff
@@ -736,7 +750,96 @@ def setPublic(self, doc, public, save=False):
+        This means any user, whether anonymous or logged in, will receive all
+        of the specified permissions. This also validates that the user attempting
+        to set the flags has permission to do so.
+
+        :param doc: The document to update permission flags on.
+        :type doc: dict
+        :param flags: Flags or set of flags to add.
+        :type flags: flag identifier, or a list/set/tuple of them
+        :param user: The user performing this action.
+        :type user:
+        :param append: Whether to append to the list or replace it.
+        :type append: bool
+        :param save: Whether to save the document to the database afterward.
+        :type save: bool
+        :param force: Set this to True to disable validation of the current
+            user's ability to set the flags.
+        :type force: bool
+        """
+        currentFlags = doc.get('publicFlags', [])
+
+        if not isinstance(flags, (list, tuple, set)):
+            flags = [flags]
+
+        if append:
+            flags = currentFlags + list(flags)
+
+        flags = set(flags) & set(PERMISSION_FLAGS.keys())
+
+        if force or user['admin']:
+            doc['publicFlags'] = list(flags)
+        else:
+            allowedFlags = []
+            for flag in flags:
+                info = PERMISSION_FLAGS[flag]
+
+                # If this is an admin-only flag, we only allow it if it's already enabled.
+                if not info['admin'] or (info['admin'] and flag in currentFlags):
+                    allowedFlags.append(flag)
+
+            doc['publicFlags'] = allowedFlags
+
+        if save:
+            doc = self.save(doc)
+
+        return doc
+
+    def _isFlagEnabled(self, doc, type, entry, flag):
+        """
+        Check whether a specific flag is enabled for a specific ACL entry in the document.
+        """
+        id = entry['id']
+        perms = doc.get('access', {})
+
+        if type == 'group':
+            return self._hasGroupAccessFlag(perms.get('groups', ()), id, flag)
+        else:  # 'user'
+            return self._hasUserAccessFlag(perms.get('users', ()), id, flag)
+
+    def _validateFlags(self, doc, user, type, entry):
+        """
+        Coerces a flag or set/list/tuple of flags into a valid form,
+        returning a list that only contains the valid flags.
+        """
+        flags = entry.get('flags', ())
+
+        if not isinstance(flags, (list, tuple, set)):
+            flags = (flags,)
+
+        flags = set(flags) & set(PERMISSION_FLAGS.keys())
+
+        if user and user['admin']:
+            return list(flags)
+
+        allowedFlags = []
+        for flag in flags:
+            info = PERMISSION_FLAGS[flag]
+
+            # If this is an admin-only flag, we only allow it if it's already enabled
+            # for this specific group or user rule
+            if info['admin'] and self._isFlagEnabled(doc, type, entry, flag):
```

Comment:
For setting flags, are there only `admin` and `non-admin` permission levels on flags?  This logic assumes such.